target_chains/ethereum/contracts/forge-test/VerificationExperiments.t.sol

```diff
@@ -0,0 +1,600 @@
+        assert(vm.payload.length == 32);
+
+        bytes32 expectedRoot = UnsafeBytesLib.toBytes32(vm.payload, 0);
+        bool validProof = isValidMerkleProof(expectedRoot, data, proof);
+        if (!validProof) revert PythErrors.InvalidArgument();
+
+        (
+            PythInternalStructs.PriceInfo memory info,
+            bytes32 priceId
+        ) = parseSingleAttestationFromBatch(data, 0, data.length);
+        uint64 latestPublishTime = latestPriceInfoPublishTime(priceId);
+
+        if (info.publishTime > latestPublishTime) {
+            setLatestPriceInfo(priceId, info);
+            emit PriceFeedUpdate(
+                priceId,
+                info.publishTime,
+                info.price,
+                info.conf
+            );
+        }
+    }
+
+    // Update a single price feed via a threshold-signed merkle proof.
+    // data is expected to be a serialized PriceAttestation
+    function updatePriceFeedsThresholdMerkle(
+        bytes memory rootSignature,
+        bytes32 rootHash,
+        bytes memory data,
+        bytes32[] memory proof
+    ) public payable {
+        if (!verifySignature(rootHash, rootSignature, thresholdPublicKey))
+            revert PythErrors.InvalidArgument();
+
+        bool validProof = isValidMerkleProof(rootHash, data, proof);
+        if (!validProof) revert PythErrors.InvalidArgument();
+
+        (
+            PythInternalStructs.PriceInfo memory info,
+            bytes32 priceId
+        ) = parseSingleAttestationFromBatch(data, 0, data.length);
+        uint64 latestPublishTime = latestPriceInfoPublishTime(priceId);
+
+        if (info.publishTime > latestPublishTime) {
+            setLatestPriceInfo(priceId, info);
+            emit PriceFeedUpdate(
+                priceId,
+                info.publishTime,
+                info.price,
+                info.conf
+            );
+        }
+    }
+
+    // Update a single price feed via a threshold-signed price update.
+    // data is expected to be a serialized PriceAttestation.
+    function updatePriceFeedsThreshold(
+        bytes memory signature,
+        bytes memory data
+    ) public payable {
+        bytes32 hash = keccak256(data);
+        if (!verifySignature(hash, signature, thresholdPublicKey))
+            revert PythErrors.InvalidArgument();
+
+        (
+            PythInternalStructs.PriceInfo memory info,
+            bytes32 priceId
+        ) = parseSingleAttestationFromBatch(data, 0, data.length);
+        uint64 latestPublishTime = latestPriceInfoPublishTime(priceId);
+
+        if (info.publishTime > latestPublishTime) {
+            setLatestPriceInfo(priceId, info);
+            emit PriceFeedUpdate(
+                priceId,
+                info.publishTime,
+                info.price,
+                info.conf
+            );
+        }
+    }
+
+    // Update a single price feed via a "native" price update (i.e., using the default ethereum tx signature for authentication).
```

Comment:
nitpik: i think for this one you need to check the message sender to have a fixed address.